### PR TITLE
RUMM-349 interceptors

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogInterceptor.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import com.datadog.android.core.internal.net.CombinedInterceptor
+import com.datadog.android.rum.RumInterceptor
+import com.datadog.android.rum.internal.net.RumRequestInterceptor
+import com.datadog.android.tracing.TracingInterceptor
+import com.datadog.android.tracing.internal.net.TracingRequestInterceptor
+import okhttp3.Interceptor
+import okhttp3.OkHttpClient
+
+/**
+ * Provides automatic integration for [OkHttpClient] by way of the [Interceptor] system.
```
Comment (mariusc83) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogInterceptor.kt:18]:
```suggestion
 * Provides automatic integration for [OkHttpClient] by using an [Interceptor] implementation.
```

File: dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import java.io.StringWriter
+import java.util.concurrent.atomic.AtomicReference
+import okhttp3.Request
+import okhttp3.Response
+
+internal class TracingRequestInterceptor : RequestInterceptor {
+
+    private val localTracerReference: AtomicReference<Tracer> = AtomicReference()
+
+    private val startedSpans = mutableMapOf<String, Span>()
```
Comment (mariusc83) [inline dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptor.kt:31]: is OkHttp using same thread for all the requests ? If not I am afraid here you are going to have a ConcurrentModificationException at some point. Can you check this by adding some tests ?

File: dd-sdk-android/src/test/kotlin/com/datadog/android/rum/internal/net/RumRequestInterceptorTest.kt [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            mockRequest,
+            fakeMethod,
+            fakeUrl,
+            emptyMap()
+        )
+        verifyNoMoreInteractions(mockRumMonitor)
+    }
+
+    @Test
+    fun `starts Rum Resource event on transformRequest within Span`(
```
Comment (mariusc83) [inline dd-sdk-android/src/test/kotlin/com/datadog/android/rum/internal/net/RumRequestInterceptorTest.kt:118]:
```suggestion
    fun `extracts the trace id as a rum attribute from the transformed request`(
```

File: dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogInterceptor.kt (hunk 1 above)
Comment (xgouchet) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogInterceptor.kt:18]: Saying it's an Interceptor implementation doesn't give any information, the signature of the class tells that. I actually reused the same javadoc template as other interceptors existing in the wild.

Comment (mariusc83) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogInterceptor.kt:18]: Sounds a bit weird for me but if it's already there like that I am fine with it.

**Pull request merged**